nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -663,359 +968,464 @@ def permuted_ols(
-    intercept_test = False
-    if n_regressors == np.unique(tested_vars).size == 1:
-        intercept_test = True
-
-    # check if confounding vars contains an intercept
-    if confounding_vars is not None:
-        # Search for all constant columns
-        constants = [
-            x
-            for x in range(confounding_vars.shape[1])
-            if np.unique(confounding_vars[:, x]).size == 1
-        ]
-
-        # check if multiple intercepts are defined across all variates
-        if (intercept_test and len(constants) == 1) or len(constants) > 1:
-            # remove all constant columns
-            confounding_vars = np.delete(confounding_vars, constants, axis=1)
-            # warn user if multiple intercepts are found
-            warnings.warn(
-                category=UserWarning,
-                message=(
-                    'Multiple columns across "confounding_vars" and/or '
-                    '"target_vars" are constant. Only one will be used '
-                    "as intercept."
-                ),
-            )
-            model_intercept = True
+    return tested_vars
+
+
+def _check_for_intercept_in_tested_var(tested_vars):
+    """Check if explanatory variates contain an intercept (constant) or not.
+
+    Parameters
+    ----------
+    tested_vars : array-like, shape=(n_samples, n_regressors)
+        Explanatory variates, fitted and tested independently from each others.
+    """
+    n_regressors = tested_vars.shape[1]
+    return n_regressors == np.unique(tested_vars).size == 1
+
+
+def _check_for_intercept_in_confounds(
+    confounding_vars, tested_var_has_intercept, model_intercept
+):
+    """Check if confounding variates contain an intercept (constant) or not.
+
+    If the tested var has an intercept, only this one is kept
+    and all other intercept are removed from the confounds.
+
+    Parameters
+    ----------
+    confounding_vars : array-like, shape=(n_samples, n_covars)
+        Confounding variates (covariates), fitted but not tested.
+
+    tested_var_has_intercept : :obj:`bool`
+        Whether the tested variate already contains an intercept or not.
+
+    model_intercept : :obj:`bool`
+        If True, a constant column will be added to the confounding variates.
+        This value may be updated if all intercepts were removed.
+
+    """
+    if confounding_vars is None:
+        test_intercept = tested_var_has_intercept
+        return confounding_vars, test_intercept, model_intercept
+
+    # Search for all constant columns
+    constants = [
+        column
+        for column in range(confounding_vars.shape[1])
+        if np.unique(confounding_vars[:, column]).size == 1
+    ]
+    # check if multiple intercepts are defined across all variates
+    if (tested_var_has_intercept and len(constants) == 1) or len(
+        constants
+    ) > 1:
+        # remove all constant columns
+        confounding_vars = np.delete(confounding_vars, constants, axis=1)
```

Comment:
doesn't this delete all constant columns (instead of leaving one) in the case where `tested_var_has_intercept is False` but `len(constants) > 1`? also what about the case where there are several regressors but one (or several) is constant?

Reply:
> took me a while to figure out the logic of this whole thing

a sign that this refactoring PR was needed!

>  I can give it a try if you think it is worth it

I think so.

and the behaviour around the intercept can maybe be sanitized a bit: ATM if the tested variates contain several columns, including a constant one, an intercept is never added to the confounds (even when testing the non-constant variates). if we think that such input makes no sense it may be worth a warning or error, or at least the variable names should reflect what is the condition: "is_intercept" rather than "has_intercept"

also, the documentation mentions a `pvals` key in the output which doesn't exist (there may be more)

although this function allows testing several variables, all the examples seem to use only one -- it might be worth seeing if testing several could be illustrated too